pkg/certificate/configuration.go

```diff
@@ -2,169 +2,453 @@ package certificate
 		}
+	}
+}
 
-		err = m.client.Update(context.TODO(), webhook)
-		if err != nil {
-			return err
+// mapWebhookToChain maps a webhook object from certificate chain data.
+func mapWebhookFromChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	clientConfigList := clientConfigMap(object.kobject)
+	for name, config := range clientConfigList {
+		serviceHostname := serviceHostname(config.Service.Name, config.Service.Namespace)
+		certificateIssue := certificateChain.CertificatesIssued[serviceHostname]
+		if certificateIssue == nil {
+			continue
 		}
-		return nil
-	})
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to update webhook CABundle")
+		caBundleName := caBundleName(object.key.String(), name)
+		caBundle := certificateIssue.CACertPEM[caBundleName]
+		if caBundle == nil {
+			continue
+		}
+		config.CABundle = caBundle
+	}
+}
+
+// mapWebhookToChain maps a secret object to certificate chain data.
+func mapSecretToChain(object *keyedObject, objects objectMap, certificateChain *chain.CertificateChainData) {
+	if object.key.NamespacedName.String() == certificateChain.CA.Name {
+		mapCASecretToChain(object, certificateChain)
+		return
+	}
+	mapServiceSecretToChain(object, certificateChain)
+}
+
+// mapWebhookToChain maps a secret object from certificate chain data.
+func mapSecretFromChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	if object.key.NamespacedName.String() == certificateChain.CA.Name {
+		mapCASecretFromChain(object, certificateChain)
+		return
+	}
+	mapServiceSecretFromChain(object, certificateChain)
+}
+
+func mapServiceSecretToChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	secret := object.kobject.(*corev1.Secret)
+	key := secret.Data[corev1.TLSPrivateKeyKey]
+	cert := secret.Data[corev1.TLSCertKey]
+	if key == nil || cert == nil {
+		return
+	}
+
+	name := serviceHostname(object.key.NamespacedName.Name, object.key.NamespacedName.Namespace)
+
+	certificateChain.CertificatesIssued[name].KeyPEM = key
+	certificateChain.CertificatesIssued[name].CertPEM = cert
+}
+
+func mapCASecretToChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	secret := object.kobject.(*corev1.Secret)
+	key := secret.Data[CAPrivateKeyKey]
+	cert := secret.Data[CACertKey]
+	if key == nil || cert == nil {
+		return
+	}
+	certificateChain.CA.KeyPEM = key
+	certificateChain.CA.CertPEM = cert
+}
+
+func mapServiceSecretFromChain(object *keyedObject, certificateChain *chain.CertificateChainData) {
+	secret := object.kobject.(*corev1.Secret)
+	name := serviceHostname(object.key.NamespacedName.Name, object.key.NamespacedName.Namespace)
+	bundle := certificateChain.CertificatesIssued[name]
+	if bundle == nil {
+		return
 	}
-	return webhook, nil
+	if secret.Data == nil {
+		secret.Data = map[string][]byte{}
+	}
+	secret.Type = corev1.SecretTypeTLS
+	secret.Data[corev1.TLSPrivateKeyKey] = bundle.KeyPEM
+	secret.Data[corev1.TLSCertKey] = bundle.CertPEM
```

Comment:
Again doing the encoding here make sense, but all depends on how important the decoupling is how feasible is that in the future we are going to use something different than the golang api structs.